dora/core/client/fs/src/main/java/alluxio/client/file/DoraFileOutStream.java

```diff
@@ -225,36 +225,45 @@ private void writeInternal(int b) throws IOException {
   }
 
   private void writeInternal(byte[] b, int off, int len) throws IOException {
-    try {
-      Preconditions.checkArgument(b != null, PreconditionMessage.ERR_WRITE_BUFFER_NULL);
-      Preconditions.checkArgument(off >= 0 && len >= 0 && len + off <= b.length,
-          PreconditionMessage.ERR_BUFFER_STATE.toString(), b.length, off, len);
-
-      if (mAlluxioStorageType.isStore() || !mClientWriteToUFSEnabled) {
-        // If client is configured to write data to worker and ask worker to write to UFS,
-        // client must send data over netty.
+    Preconditions.checkArgument(b != null, PreconditionMessage.ERR_WRITE_BUFFER_NULL);
+    Preconditions.checkArgument(off >= 0 && len >= 0 && len + off <= b.length,
+        PreconditionMessage.ERR_BUFFER_STATE.toString(), b.length, off, len);
+    if (!mClientWriteToUFSEnabled) {
+      // If client is configured to write data to worker and ask worker to write to UFS,
+      // client must send data over netty.
+      try {
         mNettyDataWriter.writeChunk(b, off, len);
-
         Metrics.BYTES_WRITTEN_ALLUXIO.inc(len);
-      }
-      if (mUnderStorageType.isSyncPersist()) {
-        if (mUnderStorageOutputStream != null) {
-          mUnderStorageOutputStream.write(b, off, len);
-          Metrics.BYTES_WRITTEN_UFS.inc(len);
+        mBytesWritten += len;
+        return;
+      } catch (IOException e) {
+        Throwable throwable = mNettyDataWriter.getPacketWriteException();
+        if (throwable != null) {
+          e.addSuppressed(throwable);
         }
+        throw e;
       }
-      mBytesWritten += len;
-    } catch (IOException e) {
-      StringBuffer exceptionMsg = new StringBuffer();
-      Throwable throwable = mNettyDataWriter.getPacketWriteException();
-      if (throwable != null) {
-        exceptionMsg.append(throwable.getMessage() + " ");
+    }
+
+    if (mAlluxioStorageType.isStore()) {
+      try {
+        mNettyDataWriter.writeChunk(b, off, len);
+        Metrics.BYTES_WRITTEN_ALLUXIO.inc(len);
+      } catch (IOException e) {
+        Throwable throwable = mNettyDataWriter.getPacketWriteException();
+        if (throwable != null) {
+          e.addSuppressed(throwable);
+        }
+        LOG.error("Failed to write data to alluxio worker. ", e);
```

Comment:
Please add comments here:
```
    //writing to worker over netty failed. But Alluxio client continues writing to UFS.
```

Reply:
Should we abort trying to write to worker for the following requests? And delete already written pages from worker cache.


The reasons are:
1. The subsequent write request might fail, too.
2. The worker will receive some data but there is already a hole (from the previous failed request). I am not sure if the worker/cache manager will handle this. I am afraid not. The cache manager on worker expects contiguous write. The write request does not have a POSITION for current data.
3. suppose for a 1MB page, the alluxio client has written 4K data at [0, 4096). And then failure for next [4096, 8192) write. And then a successful write of 4K at [8192, 12288). So, the page file on worker will have a hole, not filled with actually data. Subsequent reading from paging store will get all ZEROES for such hole. This is not correct.

Reply:
I submitted an issue https://github.com/Alluxio/alluxio/issues/18157 to make this be a future work.